pallets/lbp/src/lib.rs

```diff
@@ -1015,15 +893,108 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, BalanceOf<T>> for Pallet<T
 			.unwrap_or_else(|_| BalanceOf::<T>::zero())
 	}
 
-	/// Validate a sell trade and update pool weights if necessary
 	fn validate_sell(
 		who: &T::AccountId,
 		assets: AssetPair,
 		amount: BalanceOf<T>,
 		min_bought: BalanceOf<T>,
 		_discount: bool,
-	) -> Result<AMMTransfer<T::AccountId, AssetId, AssetPair, BalanceOf<T>>, DispatchError> {
-		Self::validate_trade(who, assets, amount, min_bought, TradeType::Sell)
+	) -> Result<AMMTransfer<T::AccountId, AssetId, AssetPair, Balance>, DispatchError> {
+
+		ensure!(!amount.is_zero(), Error::<T>::ZeroAmount);
+		ensure!(
+			T::MultiCurrency::free_balance(assets.asset_in, who) >= amount,
+			Error::<T>::InsufficientAssetBalance
+		);
+
+		let pool_id = Self::get_pair_id(assets);
+		let pool_data = <PoolData<T>>::try_get(&pool_id).map_err(|_| Error::<T>::PoolNotFound)?;
+
+		ensure!(Self::is_pool_running(&pool_data), Error::<T>::SaleIsNotRunning);
+
+		let now = T::BlockNumberProvider::current_block_number();
+		let (weight_in, weight_out) = Self::get_sorted_weight(assets.asset_in, now, &pool_data)?;
+		let asset_in_reserve = T::MultiCurrency::free_balance(assets.asset_in, &pool_id);
+		let asset_out_reserve = T::MultiCurrency::free_balance(assets.asset_out, &pool_id);
+
+		ensure!(
+			amount <= asset_in_reserve.checked_div(MAX_IN_RATIO).ok_or(Error::<T>::Overflow)?,
+			Error::<T>::MaxInRatioExceeded
+		);
+		
+		// LBP fee asset is always accumulated asset
+		let fee_asset = pool_data.assets.0;
+
+		// Accumulated asset is sold (in) to the pool for distributed asset (out)
+		// Take accumulated asset (in) sans fee from the seller and add to pool
+		// Take distributed asset (out) and send to seller 
+		// Take fee from the seller and add to fee repay pool
+		// Pool bears repay fee
+		if assets.asset_in == pool_data.assets.0 {
+			let fee = Self::calculate_fees(&pool_data, amount)?;
+			let amount_without_fee = amount.checked_sub(fee).ok_or(Error::<T>::Overflow)?;
```

Comment:
would move fee calc right before return, to show clearly that this value is not used sooner